runtime/near-evm-runner/src/evm_state.rs

```diff
@@ -0,0 +1,533 @@
+        Ok(account.balance.into())
+    }
+
+    fn nonce_of(&self, address: &Address) -> Result<U256> {
+        let account = self.get_account(address)?.unwrap_or_default();
+        Ok(account.nonce.into())
+    }
+
+    fn set_nonce(&mut self, address: &Address, nonce: U256) -> Result<Option<U256>> {
+        let mut account = self.get_account(address)?.unwrap_or_default();
+        account.nonce = nonce;
+        self.set_account(address, &account)?;
+        Ok(Some(account.nonce))
+    }
+
+    fn set_balance(&mut self, address: &Address, balance: U256) -> Result<()> {
+        let mut account = self.get_account(address)?.unwrap_or_default();
+        account.balance = balance;
+        self.set_account(address, &account)
+    }
+
+    fn next_nonce(&mut self, address: &Address) -> Result<U256> {
+        let mut account = self.get_account(address)?.unwrap_or_default();
+        let nonce = account.nonce;
+        account.nonce += U256::from(1);
+        self.set_account(address, &account)?;
+        Ok(nonce)
+    }
+
+    fn _read_contract_storage(&self, key: [u8; 52]) -> Result<Option<[u8; 32]>>;
+    fn read_contract_storage(&self, address: &Address, key: [u8; 32]) -> Result<Option<[u8; 32]>> {
+        self._read_contract_storage(utils::internal_storage_key(address, key))
+    }
+
+    fn _set_contract_storage(&mut self, key: [u8; 52], value: [u8; 32]) -> Result<()>;
+
+    fn set_contract_storage(
+        &mut self,
+        address: &Address,
+        key: [u8; 32],
+        value: [u8; 32],
+    ) -> Result<()> {
+        self._set_contract_storage(utils::internal_storage_key(address, key), value)
+    }
+
+    fn commit_changes(&mut self, other: &StateStore) -> Result<()>;
+
+    fn add_balance(&mut self, address: &Address, incr: U256) -> Result<()> {
+        let mut account = self.get_account(address)?.unwrap_or_default();
+        account.balance = account
+            .balance
+            .checked_add(incr)
+            .ok_or_else(|| VMLogicError::EvmError(EvmError::IntegerOverflow))?;
+        self.set_account(address, &account)
+    }
+
+    fn sub_balance(&mut self, address: &Address, decr: U256) -> Result<()> {
+        let mut account = self.get_account(address)?.unwrap_or_default();
+        account.balance = account
+            .balance
+            .checked_sub(decr)
+            .ok_or_else(|| VMLogicError::EvmError(EvmError::InsufficientFunds))?;
+        self.set_account(address, &account)
+    }
+
+    fn transfer_balance(
+        &mut self,
+        sender: &Address,
+        recipient: &Address,
+        amnt: U256,
+    ) -> Result<()> {
+        self.sub_balance(sender, amnt)?;
+        self.add_balance(recipient, amnt)
+    }
+
+    fn recreate(&mut self, address: [u8; 20]);
+}
+
+#[derive(Default, Debug)]
+pub struct StateStore {
+    pub code: HashMap<RawAddress, Vec<u8>>,
+    pub accounts: HashMap<RawAddress, EvmAccount>,
+    pub storages: BTreeMap<Vec<u8>, [u8; 32]>,
```

Comment:
This looks like will cost a lot of memory, should we use rocksdb here.

Reply:
This is temporary data structure of overlay changes, similar how it's in StoreUpdate in our storage

Reply:
Yeah just realized from runtime/near-evm-runner/src/lib.rs